staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_interval.go

```diff
@@ -133,9 +134,22 @@ func (s sortableWatchCacheEvents) Swap(i, j int) {
 // returned by Next() need to be events from a List() done on the underlying store of
 // the watch cache.
 // The items returned in the interval will be sorted by Key.
-func newCacheIntervalFromStore(resourceVersion uint64, store cache.Indexer, getAttrsFunc attrFunc) (*watchCacheInterval, error) {
+func newCacheIntervalFromStore(resourceVersion uint64, store cache.Indexer, getAttrsFunc attrFunc, key string, predicate *storage.SelectionPredicate) (*watchCacheInterval, error) {
 	buffer := &watchCacheIntervalBuffer{}
-	allItems := store.List()
+	var allItems []interface{}
+
+	if _, ok := predicate.MatchesSingle(); ok {
+		item, exists, err := store.GetByKey(key)
```

Comment:
OK - so it seems we're setting key correctly here - so this would work: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L382-L385